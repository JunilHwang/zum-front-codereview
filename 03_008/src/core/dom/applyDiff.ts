const differentValue = (node1: HTMLInputElement, node2: HTMLInputElement) => {
  if (node1.value !== node2.value) {
    return true;
  }

  if (node1.checked !== node2.checked) {
    return true;
  }

  return false;
};

const isNodeChanged = (node1: Element, node2: Element) => {
  const n1Attributes = node1.attributes;
  const n2Attributes = node2.attributes;
  // 노드의 어트리뷰트 갯수가 다르다면 변화된 노드
  if (n1Attributes.length !== n2Attributes.length) {
    return true;
  }

  const differentAttribute = Array.from(n1Attributes).find(attribute => {
    const { name } = attribute;
    const attribute1 = node1.getAttribute(name);
    const attribute2 = node2.getAttribute(name);

    return attribute1 !== attribute2;
  });

  // 어트리뷰트의 값이 다른게 하나라도 있다면 변화된것
  if (differentAttribute) {
    return true;
  }

  // 만약 value프로퍼티의 값이 바뀌었다면 변화된것
  if (differentValue(node1 as HTMLInputElement, node2 as HTMLInputElement)) {
    return true;
  }

  // 하위노드의 갯수가 다르거나 텍스트노드의 값이 다를 경우 변화환것
  if (
    node1.children.length === 0 &&
    node2.children.length === 0 &&
    node1.textContent !== node2.textContent
  ) {
    return true;
  }

  // 위의 경우가 다 틀리다면 변화하지 않은것으로 판단하고 리렌더링을 하지 않는다.
  return false;
};

export const applyDiff = (parentNode: Element, realNode: Element, virtualNode: Element) => {
  // 실제 DOM에는 있지만 상태 변화가 적용된 가상 DOM에 node가 존재하지 않는다면 실제 DOM에서 삭제처리
  if (realNode && !virtualNode) {
    realNode.remove();
    return;
  }

  // 실제 돔에는 없지만 상태 변화가 적용된 가상돔에는 node가 존재한다면 생성
  if (!realNode && virtualNode) {
    parentNode.appendChild(virtualNode);
    return;
  }

  // 어트리뷰트, children등이 변경된지 판단하고 실제 DOM의 node를 replace처리
  if (isNodeChanged(virtualNode, realNode)) {
    realNode.replaceWith(virtualNode);
    return;
  }

  // children들만 배열로 추출
  const realChildren = Array.from(realNode.children);
  const virtualChildren = Array.from(virtualNode.children);

  // 상태가 적용된 가상 DOM은 children의 갯수가 다를 수 있다. 최대 children의 갯수를 구해준다.
  const max = Math.max(realChildren.length, virtualChildren.length);
  // 재귀적으로 가장 깊은 node까지 탐색한다
  for (let i = 0; i < max; i++) {
    // 여기서 첫번째 인수로 전달된 realNode는 상태에 따라 remove, append, replace 작업이 완료된 node이다. 단순히 appendChild를 하기 위해 전달한다.
    applyDiff(realNode, realChildren[i], virtualChildren[i]);
  }
};
