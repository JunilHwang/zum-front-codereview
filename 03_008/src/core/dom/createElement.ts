export const createElement = (
  DOMstring: string,
  children?: Array<Element | null | false | undefined>
) => {
  const $template = document.createElement('template');
  $template.innerHTML = DOMstring;

  const $element = $template.content.firstElementChild?.cloneNode(true) as Element;

  if (children) {
    children.forEach(child => {
      if (!child) return;
      $element.appendChild(child);
    });
  }

  return $element;
};
