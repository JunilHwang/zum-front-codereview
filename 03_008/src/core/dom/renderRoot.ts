const renderRoot = ($root: Element, $app: Element) => {
  const $newRoot = $root.cloneNode(false) as Element;
  // 상태를 전역으로 참조하고 있기때문에 $app은 모듈화 되어있는 컴포넌트들이 변경된 상태를 가지고 본인의 모양을 가지게 된다.
  $newRoot.appendChild($app);

  return $newRoot;
};

export default renderRoot;
