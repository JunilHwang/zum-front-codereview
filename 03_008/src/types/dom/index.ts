export type commonProps<P> = P & {
  children?: Element[] | undefined;
  className?: string;
  textContent?: string;
  id?: string;
};

export interface FunctionComponent<P = {}> {
  (props: commonProps<P>): Element;
}
