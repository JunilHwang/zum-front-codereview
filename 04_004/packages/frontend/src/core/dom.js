import { instantiateTree } from './reconciler';

export class Component {
  constructor(props = {}) {
    this.props = props;
    this.state = {};
  }

  setState(partialState) {
    this.state = {
      ...this.state,
      ...partialState,
    };
  }
}
Component.prototype.isComponent = {};

/* istanbul ignore next */
function warnIfKeyNotExists(children) {
  if (
    children.some(child => typeof child.props.key === 'undefined') ||
    new Set(children.map(child => child.props.key)).size !== children.length
  ) {
    console.warn('children of list should have unique keys');
  }
}

function childToString(child) {
  return ['object', 'function'].includes(typeof child) ?
    child :
    String(child);
}

function wrapChildren(children) {
  /* istanbul ignore next */
  if (__DEV__ && children.some(child => Array.isArray(child))) {
    children.forEach(warnIfKeyNotExists);
  }
  return children.flat().map(childToString);
}

export function createElement(type, props, ...children) {
  return {
    type,
    props: Object
      .assign(
        {},
        props,
        children.length > 0 &&
        { children: wrapChildren(children) } ||
        { children }),
  };
}

export function createRef() {
  return {current: null};
}

function clearContainer(container) {
  container.childNodes
    .forEach(container.removeChild.bind(container));
}

export function render(element, container) {
  const instance = instantiateTree(element);
  if (container._mountedInstanceTree) {
    const mounted = container._mountedInstanceTree;
    mounted.diff(element);
    return;
  }
  clearContainer(container);

  const rootNode = instance.mount();
  container._mountedInstanceTree = instance;
  container.appendChild(rootNode);
}
