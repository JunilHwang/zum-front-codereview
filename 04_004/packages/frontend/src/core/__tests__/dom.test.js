/**
 * @jest-environment jsdom
 */
import { fireEvent, getByText, queryByText } from '@testing-library/dom';
import { createElement as h, render, Component, createRef } from '../dom';

let container;
beforeEach(() => {
  container = document.createElement('div');
  jest.useFakeTimers();
});

afterEach(() => {
  container.innerHTML = '';
  jest.useRealTimers();
});

const act = callback => {
  callback();
  jest.runAllTimers();
};

describe('render', () => {
  test('hello world 렌더링', () => {
    render(h('h1', null, 'hello world'), container);
    expect(container).toMatchSnapshot();
  });

  test('hello world 컴포넌트 렌더링', () => {
    class App extends Component {
      render() {
        return h('h1', null, 'hello world');
      }
    }
    render(h(App), container);
    expect(container).toMatchSnapshot();
  });

  describe('변경사항 적용', () => {
    test('텍스트 노드 변경', () => {
      render(h('h1', null, 'hello world'), container);
      const header = getByText(container, 'hello world');
      act(() => {
        render(h('h1', null, 'foobar'), container);
      });
      expect(header.textContent).toBe('foobar');
    });

    test('속성 변경', () => {
      render(h('a', {href: 'https://www.google.com/'}, 'go!'), container);
      const link = container.querySelector('a[href="https://www.google.com/"]');
      act(() => {
        render(h('a', {href: 'https://github.com/'}, 'go!'), container);
      });
      expect(link.href).toBe('https://github.com/');
    });

    test('boolean값 속성', () => {
      render(h('button', {type: 'button', disabled: true}, 'click'), container);
      const button = container.querySelector('button[disabled]');
      expect(button).not.toBeNull();
      act(() => {
        render(h('button', {type: 'button', disabled: false}, 'click'), container);
      });
      expect(button.getAttribute('disabled')).toBeNull();
    });

    test('이벤트리스너 제어', () => {
      const mock = jest.fn();
      render(h('button', {type: 'button', onClick: mock}, 'click'), container);
      act(() => {
        fireEvent.click(container.querySelector('button'));
      });
      expect(mock).toBeCalledTimes(1);
      render(h('button', {type: 'button'}, 'click'), container);
      act(() => {
        fireEvent.click(container.querySelector('button'));
      });
      expect(mock).toBeCalledTimes(1);
      render(h('button', {type: 'button', onClick: mock}, 'click'), container);
      act(() => {
        fireEvent.click(container.querySelector('button'));
      });
      expect(mock).toBeCalledTimes(2);
    });
  });
});

describe('Component', () => {
  describe('생명주기 메서드', () => {
    test('componentDidMount', () => {
      class App extends Component {
        componentDidMount() {
          this.props.mock();
        }

        render() {
          return null;
        }
      }
      const mock = jest.fn();
      act(() => {
        render(h(App, {mock}), container);
      });
      expect(mock).toBeCalled();
    });

    test('componentWillUnmount', () => {
      class Children extends Component {
        componentWillUnmount() {
          this.props.mock();
        }

        render() {
          return null;
        }
      }

      class App extends Component {
        render() {
          return this.props.show && h(Children, {mock: this.props.mock});
        }
      }
      const mock = jest.fn();
      act(() => {
        render(h(App, {show: true, mock}), container);
      });
      act(() => {
        render(h(App, {show: false, mock}), container);
      });
      expect(mock).toBeCalled();
    });
  });

  describe('props', () => {
    test('props값을 이용한 렌더링', () => {
      class App extends Component {
        render() {
          return h('h1', null, this.props.message);
        }
      }
      render(h(App, {message: 'hi'}), container);
      expect(queryByText(container, 'hi')).not.toBeNull();
    });

    test('onClick 이벤트리스너', () => {
      class Counter extends Component {
        constructor(props) {
          super(props);

          this.state = {count: 0};
        }

        increment() {
          this.setState({
            count: this.state.count + 1,
          });
        }

        render() {
          return h('div', null,
            h('span', null, this.state.count),
            h('button', {onClick: this.increment.bind(this)}, 'click')
          );
        }
      }
      act(() => {
        render(h(Counter), container);
      });
      const button = getByText(container, 'click');

      act(() => {
        fireEvent.click(button);
      });
      expect(queryByText(container, '1')).not.toBeNull();
    });
  });
});

test('createRef', () => {
  const buttonRef = createRef();
  expect(buttonRef.current).toBeNull();
  render(h('button', {ref: buttonRef}, 'click'), container);
  expect(buttonRef.current).not.toBeNull();
});