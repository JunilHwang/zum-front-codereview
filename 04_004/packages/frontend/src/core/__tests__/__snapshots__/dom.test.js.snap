// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render hello world 렌더링 1`] = `
<div>
  <h1>
    hello world
  </h1>
</div>
`;

exports[`render hello world 컴포넌트 렌더링 1`] = `
<div>
  <h1>
    hello world
  </h1>
</div>
`;
